Woodoku/src/game.py

```diff
@@ -1,17 +1,49 @@
 import random
 from typing import List
 
+import yaml
+
 from entity.woodoku_board import WoodokuBoard
 from entity.woodoku_shape import WoodokuShape
 from ui.command_line_ui import CommandLineUI
 from ui.ui_interface import UIInterface
 
-CONFIG_FILE = ''
+CONFIG_FILE = './config.yaml'
 NUM_SHAPES = 3
 
 
 def read_shapes_from_file(filepath: str) -> List[WoodokuShape]:
-    pass
+    raw_shapes = []
+    with open(filepath) as config:
+        raw_shapes_list = yaml.safe_load(config)['raw_shapes']
+        for row in raw_shapes_list:
+            tuple_row = [tuple(x) for x in row]
+            raw_shapes.append(WoodokuShape(tuple_row))
+
+    return raw_shapes
+
+
+def rotate_all_shapes(raw_shapes: List[WoodokuShape]) -> List[WoodokuShape]:
+    """
+    Rotate each shape in <raw_shapes> for three times, then gather all of them.
+
+    Args:
+        raw_shapes: The default shape without any rotation.
+
+    Returns:
+        The original shapes and all the rotated shapes (non-repeated).
+    """
+    shapes = set()
+
+    # All the raw shape itself was added first, then rotates three times and add to set after each rotate.
+    # Duplicates are eliminated by set operation.
+    for shape in raw_shapes:
+        shapes.add(shape)
+        for _ in range(3):
+            new_shape = shape.rotate()
```

Comment:
`shape.rotate` require some implementation considerations. This is linked to 
#34 and the boarder of a shape not being defined. 

Now we decided to implement shapes with the top left corner as (0,0)
Presumably, at least a block is occupied in the first row or first column. 
e.g. 

|  	| ■ 	|  	|
|---	|:---:	|---	|
| ■ 	| ■ 	| ■ 	|

has blocks occupied at (0, 1) and (1, 0) but not at (0,0)

the problem we run into is when rotating 3 times we need a smart mechanism to adhere to the above rule. 

if we rotate the shape above clockwise we will have the first column empty.
which violates the above rule and causes problems when placing shape.

|  	|  	| ■ 	|
|---	|:---:	|---	|
|  	| ■ 	| ■ 	|
|  	|  	| ■ 	|

a simple example of a problem is placing the rotated shape at (0,0), you simply can't place any blocks on the first column

Reply:
A second problem we face is invalid block as @YANLING-H mentioned and created #34. 

when we place a 3x3 shape the highest row and column index to place that shape can only be at (78,78) and not at (80,80), otherwise we have an invalid block either when checking or at placing a block.

Reply:
I'll move my suggested solution to #34

Reply:
> if we rotate the shape above clockwise we will have the first column empty.
> which violates the above rule and causes problems when placing shape.

Yeah I considered that. I intended to implement that in the `rotate()` method.

Reply:
> when we place a 3x3 shape the highest row and column index to place that shape can only be at (78,78) and not at (80,80), otherwise we have an invalid block either when checking or at placing a block.

I am not sure what this means. Could you elaborate?

Reply:
alright, use handle the exception to solve this